<!DOCTYPE html>
<html lang="zh-CN">
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- 上述3个meta标签*必须*放在最前面，任何其他内容都*必须*跟随其后！ -->
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="icon" href="favicon.ico">

    <title>生命之美结题报告</title>
<link rel="stylesheet" href="assets/css/jquery-ui.css">
    <!-- Bootstrap core CSS -->
    <link href="assets/css/bootstrap.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="assets/css/lightbox.css" rel="stylesheet" />
    <link href="assets/css/markdown.css" rel="stylesheet">
    
    <link href="assets/css/report.css" rel="stylesheet">

    <!-- Just for debugging purposes. Don't actually copy these 2 lines! -->
    <!--[if lt IE 9]>
    <script src="assets/js/ie8-responsive-file-warning.js"></script><![endif]-->
    <script src="ie-emulation-modes-warning.js"></script>

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
    <script src="assets/js/html5shiv.min.js"></script>
    <script src="assets/js/respond.min.js"></script>
    <![endif]-->


    <link rel="stylesheet" href="assets/css/zTreeStyle.css" type="text/css">

  

</head>

<body>
<div id="menubar" class="noprint"><a href="

<div id="container" class="container">


    <!-- Main component for a primary marketing message or call to action -->
    <div id="page-top" class="page noprintborder">


        <div class="pageheader-first" style="text-align:center;">
            <img src="assets/img/header3.png" height="60" width="327">
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>
        <div class="title1">结题报告</div>
        <div class="title2">转录组分析结题报告</div>
        <div class="title3">技术委托服务项目</div>
        <div class="info">
            <div><span class="info1">委托方（甲方）：</span><span class="info2">[ablife:institutions]</span></div>
            <div><span class="info1">受托方（乙方）：</span><span class="info2">武汉生命之美科技有限公司</span></div>
            <div><span class="info1">报告日期：</span><span class="info2">[ablife:reportdate]</span></div>
        </div>
        <div class="timepoint">
            <p class="timetitle">项目执行的重要时间点：</p>
            
            <p>样品收讫：2014年3月4日</p>

            
        </div>
        <div class="foot">武汉生命之美科技有限公司，武汉，中国</div>

        <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>
    <!-- /pageend -->


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

    <h1>Part Ⅰ RNA-seq数据分析方案流程</h1>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="assets/public/RNA-seq_workflow.jpg" data-lightbox="g1" data-title="Figure 1.1．RNA-seq workflow" class="thumbnail"><img src="assets/public/RNA-seq_workflow.jpg" alt="Figure 1.1．RNA-seq workflow" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h1>Part Ⅱ 建库与测序信息</h1>

<h2>2.1方法描述</h2>

<p>Total RNA was treated with RQ1 DNase (promega) to remove DNA. The quality and quantity of the purified RNA were determined by measuring the absorbance at 260nm/280nm (A260/A280) using smartspec plus (BioRad). RNA integrity was further verified by 1.5% Agarose gel electrophoresis.</p>

<p>For each sample, 10μg of total RNA was used for RNA-seq library preparation. Polyadenylated mRNAs were purified and concentrated with oligo(dT)-conjugated magnetic beads (invitrogen) before used for directional RNA-seq library preparation. Purified mRNAs were iron fragmented at 95℃ followed by end repair and 5' adaptor ligation. Then reverse transcription was performed with RT primer harboring 3' adaptor sequence and randomized hexamer. The cDNAs were purified and amplified and PCR products corresponding to 200-500 bps were purified, quantified and stored at -80℃ until used for sequencing.</p>

<p>For high-throughput sequencing, the libraries were prepared following the manufacturer's instructions and applied to illumina NextSeq 500 system for 151 nt pair-end sequencing by ABlife. Inc (Wuhan, China).</p>

       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h2>2.2 建库部分结果展示</h2>

<h3>2.2.1 电泳检测</h3>

	<div class="row"><div class="col-xs-6 col-md-6 center-block" style="float:none;"><a href="RNA-seq/Sup0_Library/电泳检测.jpg" data-lightbox="g2" data-title="图 2.1．电泳检测" class="thumbnail"><img src="RNA-seq/Sup0_Library/电泳检测.jpg" alt="图 2.1．电泳检测" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>2.2.2 mRNA/ncRNA捕获</h3>

	<div class="row"><div class="col-xs-6 col-md-6 center-block" style="float:none;"><a href="RNA-seq/Sup0_Library/mRNA捕获后检测.jpg" data-lightbox="g3" data-title="图 2.2．mRNA捕获后检测" class="thumbnail"><img src="RNA-seq/Sup0_Library/mRNA捕获后检测.jpg" alt="图 2.2．mRNA捕获后检测" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>2.2.3 cDNA文库</h3>

	<div class="row"><div class="col-xs-6 col-md-6 center-block" style="float:none;"><a href="RNA-seq/Sup0_Library/cDNA片段.jpg" data-lightbox="g4" data-title="图 2.3．PCR duplication level统计" class="thumbnail"><img src="RNA-seq/Sup0_Library/cDNA片段.jpg" alt="图 2.3．PCR duplication level统计" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h2>2.3 样本信息</h2>

<div><p class="table_info">Table2.1. Sample description</p></div>
<div><p class="table_info">(表2.1：样品信息)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample ID</th><th>Name</th><th>Description
</th></tr>
</thead><tbody><tr><td>ABL2014-04012-1</td><td>Normal_5</td><td>5-0对照
</td></tr>
<tr><td>ABL2014-04012-2</td><td>Acute_11</td><td>11 慢性急变
</td></tr>
<tr><td>ABL2014-04012-3</td><td>Acute_13</td><td>13 慢性急变
</td></tr>
<tr><td>ABL2014-04012-4</td><td>Acute_14</td><td>14 慢性急变
</td></tr>
<tr><td>ABL2014-04012-5</td><td>Acute_15</td><td>15-1 实验B
</td></tr>
<tr><td>ABL2014-04012-6</td><td>Normal_3</td><td>3 对照
</td></tr>
<tr><td>ABL2014-04012-7</td><td>Chronic_8</td><td>8 实验A
</td></tr>
<tr><td>ABL2014-04012-8</td><td>Chronic_6</td><td>6 实验A
</td></tr>
<tr><td>ABL2014-04012-9</td><td>Chronic_7</td><td>7 实验A
</td></tr>
<tr><td>ABL2014-04012-10</td><td>Chronic_9</td><td>9 实验A
</td></tr>
<tr><td>ABL2014-04012-11</td><td>Chronic_10</td><td>10-1 实验A
</td></tr>
<tr><td>ABL2014-04012-12</td><td>Acute_2</td><td>12 慢性急变
</td></tr>
<tr><td>ABL2014-04012-13</td><td>Normal_2</td><td>2 对照
</td></tr>
<tr><td>ABL2014-04012-14</td><td>Normal_4</td><td>4 对照
</td></tr>
<tr><td>ABL2014-04012-15</td><td>Normal_1</td><td>1 对照
</td></tr>
</tbody></table></div></div>
<h2>2.4 测序数据</h2>

<p>本项目对甲方提供的样本成功进行了建库和高通量测序。建库结果使用illumina Nextseq 500测序平台进行双端（paired-end）测序，获得了高质量的转录组数据。</p>

<p><strong>方法描述：</strong></p>

<p>Clean reads指剔除了adapter自连及低质量的碱基和reads后的有效reads。</p>

<p><strong>有效序列（clean reads）提取策略说明：</strong></p>

<p><strong>1端clean策略：</strong></p>

<ol>
<li><p>截掉前面3个碱基；</p></li>
<li><p>截掉mRNA的adapter，并保留长度17以上的reads；</p></li>
<li><p>截掉末尾的一个碱基；</p></li>
<li><p>去掉质量低于20的碱基，保留长度为16；</p></li>
<li><p>去掉30%碱基质量低于20的reads；</p></li>
<li><p>discard NN，discard polyG和polyA，保留长度为16以上的reads；</p></li>
</ol>

<p><strong>2端clean策略：</strong></p>

<ol>
<li><p>截掉mRNA的adapter，并保留长度20以上的reads；</p></li>
<li><p>截掉末尾的4个碱基；</p></li>
<li><p>去掉质量低于20的碱基，保留长度为16；</p></li>
<li><p>去掉30%碱基质量低于20的reads；</p></li>
<li><p>discard NN，discard polyG和polyA，保留长度为16以上的reads。</p></li>
</ol>

       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<p><strong>结果展示:</strong></p>

<div><p class="table_info">Table 2.2. Obtain the high-quality clean reads</p></div>
<div><p class="table_info">(表2.2：获得高质量的有效序列)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample </th><th>Raw data</th><th>Clean reads</th><th>clean ratio</th><th>uniqtag
</th></tr>
</thead><tbody><tr><td>Normal_5</td><td>55935904</td><td>34381801</td><td>61.47%</td><td>29984069<br>(87.21%)
</td></tr>
<tr><td>Acute_11</td><td>59933534</td><td>37648510</td><td>62.82%</td><td>31198307<br>(82.87%)
</td></tr>
<tr><td>Acute_13</td><td>61600980</td><td>46332397</td><td>75.21%</td><td>32301230<br>(69.72%)
</td></tr>
<tr><td>Acute_14</td><td>58965068</td><td>41068944</td><td>69.65%</td><td>26106893<br>(63.57%)
</td></tr>
<tr><td>Acute_15</td><td>61463922</td><td>39631158</td><td>64.48%</td><td>25943149<br>(65.46%)
</td></tr>
<tr><td>Normal_3</td><td>41506308</td><td>33467594</td><td>80.63%</td><td>25142626<br>(75.13%)
</td></tr>
<tr><td>Chronic_8</td><td>45152864</td><td>33730612</td><td>74.70%</td><td>25463607<br>(75.49%)
</td></tr>
<tr><td>Chronic_6</td><td>45491400</td><td>36597709</td><td>80.45%</td><td>26039596<br>(71.15%)
</td></tr>
<tr><td>Chronic_7</td><td>42037788</td><td>32531520</td><td>77.39%</td><td>23716985<br>(72.90%)
</td></tr>
<tr><td>Chronic_9</td><td>57552342</td><td>41135625</td><td>71.48%</td><td>25027170<br>(60.84%)
</td></tr>
<tr><td>Chronic_10</td><td>57978028</td><td>36103120</td><td>62.27%</td><td>19130283<br>(52.99%)
</td></tr>
<tr><td>Acute_12</td><td>64615906</td><td>48212364</td><td>74.61%</td><td>28851925<br>(59.84%)
</td></tr>
<tr><td>Normal_2</td><td>42185520</td><td>32961659</td><td>78.14%</td><td>25326361<br>(76.84%)
</td></tr>
<tr><td>Normal_4</td><td>42099614</td><td>34261878</td><td>81.38%</td><td>25171840<br>(73.47%)
</td></tr>
<tr><td>Normal_1</td><td>33964190</td><td>27064427</td><td>79.69%</td><td>20998625<br>(77.59%)
</td></tr>
</tbody></table></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h1>Part Ⅲ 转录组测序项目结果-基础分析</h1>

<h2>3.1 RNA-seq高通量测序数据质量分析</h2>

<h3>3.1.1 Clean reads测序质量分析</h3>

<p><strong>方法描述</strong></p>

<p>Note: Sequencing quality distribution：</p>

<p>y- axis represents the sequencing quality score , it was calculated through the following formula ：</p>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="assets/public/sequencing quality score formula.jpg" data-lightbox="g5" data-title="Figure 3.1．sequencing quality score formula." class="thumbnail"><img src="assets/public/sequencing quality score formula.jpg" alt="Figure 3.1．sequencing quality score formula." /></a></div></div>
<p>Where p is the probability that the corresponding base call is incorrect, x-axis represents each position of the read(bp), when the Q_(solexa-prior to v.1.3)= 28，the p=0.001582385</p>

<p>Sequence content distribution:</p>

<p>y- axis represents the percentage of each kind of nucleotide, x-axis represents each position of the read(bp)</p>

       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<p><strong>结果展示：</strong></p>

		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g6" aria-expanded="true" aria-controls="g6">
		          per_base_quality
		        </a>
		      </div>
		    </div>
		    <div id="g6" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">chronic_8.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">chronic_9.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">normal_5.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">acute_13.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">normal_2.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">normal_3.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">chronic_7.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">acute_14.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">acute_15.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">chronic_10.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">acute_12.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">normal_4.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">normal_1.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">acute_11.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/per_base_quality.png" data-lightbox="g6" data-title="per_base_quality" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/per_base_quality.png" alt="per_base_quality" /><div class="caption">chronic_6.clean_fq_clipper.fq_fastqc</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.2 有效长度统计</h3>

<p><strong>方法描述：</strong></p>

<p>去掉index序列、建库平衡用随机碱基及截取掉后面低质量的碱基后，我们用获得的clean reads进行有效长度分析。</p>

<p><strong>结果展示：</strong></p>

		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g7" aria-expanded="true" aria-controls="g7">
		          sequence_length_distribution
		        </a>
		      </div>
		    </div>
		    <div id="g7" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">chronic_8.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">chronic_9.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">normal_5.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">acute_13.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">normal_2.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">normal_3.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">chronic_7.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">acute_14.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">acute_15.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">chronic_10.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">acute_12.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">normal_4.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">normal_1.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">acute_11.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" data-lightbox="g7" data-title="sequence_length_distribution" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/sequence_length_distribution.png" alt="sequence_length_distribution" /><div class="caption">chronic_6.clean_fq_clipper.fq_fastqc</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.3 cDNA库的代表性分析</h3>

<p><strong>方法描述：</strong></p>

<p>在cDNA文库构建的过程中对捕获的mRNA/ncRNA 进行随机片段化，然后加接头和RT-PCR。如果PCR扩增过度，cDNA文库中会含有大量序列完全相同的reads。序列完全相同的reads的来源PCR过度扩增外，还有高丰度RNA在片段化的时候产生的大量5’end相同的片段。为了保证cDNA库更好代表原始mRNA/ncRNA的表达量，本公司在建库中对PCR扩增循环有严格控制。</p>

<p>PCR duplication level计算方法为：从测序数据中随机挑选20万reads作为Total Reads，按照如下公式进行计算：</p>

<p><center></p>

<p><strong>PCR duplication level=Duplication Reads/Total Reads</strong></p>

<p></center></p>

<p><strong>结果展示：</strong></p>

		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g8" aria-expanded="true" aria-controls="g8">
		          duplication_levels
		        </a>
		      </div>
		    </div>
		    <div id="g8" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_8.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">chronic_8.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_9.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">chronic_9.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_5.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">normal_5.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_13.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">acute_13.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_2.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">normal_2.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_3.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">normal_3.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_7.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">chronic_7.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_14.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">acute_14.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_15.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">acute_15.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_10.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">chronic_10.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_12.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">acute_12.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_4.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">normal_4.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/normal_1.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">normal_1.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/acute_11.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">acute_11.clean_fq_clipper.fq_fastqc</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/duplication_levels.png" data-lightbox="g8" data-title="duplication_levels" class="thumbnail"><img src="RNA-seq/Sup1_Quality_control/chronic_6.clean_fq_clipper.fq_fastqc/duplication_levels.png" alt="duplication_levels" /><div class="caption">chronic_6.clean_fq_clipper.fq_fastqc</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h2>3.2 全基因组定位分析</h2>

<h3>3.2.1 参考基因组信息</h3>

<div><p class="table_info">Table3.1.Length distribution across Genomic Regions</p></div>
<div><p class="table_info">(表3.1：参考基因组信息.)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample</th><th>CDS</th><th>five_prime_UTR</th><th>three_prime_UTR</th><th>Intergenic</th><th>intron</th><th>Noncoding exon
</th></tr>
</thead><tbody><tr><td>human</td><td>35293030.00<br>(1.14%)</td><td>4533072.00<br>(0.15%)</td><td>24827485.00<br>(0.80%)</td><td>1808026205.00<br>(58.40%)</td><td>1211395392.00<br>(39.13%)</td><td>11618799.00<br>(0.38%)</td></tr>
</tbody></table></div></div>


	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="RNA-seq/Reference_genome/Length distribution across Genomic Regions.png" data-lightbox="g9" data-title="Figure 3.2．Length distribution across Genomic Regions" class="thumbnail"><img src="RNA-seq/Reference_genome/Length distribution across Genomic Regions.png" alt="Figure 3.2．Length distribution across Genomic Regions" /></a></div></div>
<h3>3.2.2 Reads比对到参考基因组结果</h3>

<p><strong>方法描述：</strong></p>

<p>在这部分分析中，我们将有效测序数据（clean reads）比对到参考基因组上。</p>

<p><strong>结果展示：</strong></p>

<div><p class="table_info">Table3.2. Mapping of clean reads on the reference genome</p></div>
<div><p class="table_info">(表3.2：有效测序数据比对到参考基因组.)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample</th><th>Input reads</th><th>Total mapped</th><th>Unique mapped</th><th>Multiple mapped
</th></tr>
</thead><tbody><tr><td>acute_11</td><td>37648510</td><td>24774704<br>(65.81%)</td><td>21533451<br>(86.92%)</td><td>3241253<br>(13.08%)
</td></tr>
<tr><td>acute_12</td><td>48212364</td><td>38372355<br>(79.59%)</td><td>34832862<br>(90.78%)</td><td>3539493<br>(9.22%)
</td></tr>
<tr><td>acute_13</td><td>46332397</td><td>36457880<br>(78.69%)</td><td>32751411<br>(89.83%)</td><td>3706469<br>(10.17%)
</td></tr>
<tr><td>acute_14</td><td>41068944</td><td>29524310<br>(71.89%)</td><td>25727844<br>(87.14%)</td><td>3796466<br>(12.86%)
</td></tr>
<tr><td>acute_15</td><td>39631158</td><td>28493677<br>(71.90%)</td><td>25047782<br>(87.91%)</td><td>3445895<br>(12.09%)
</td></tr>
<tr><td>chronic_10</td><td>36103120</td><td>26025612<br>(72.09%)</td><td>22933734<br>(88.12%)</td><td>3091878<br>(11.88%)
</td></tr>
<tr><td>chronic_6</td><td>36597709</td><td>29126739<br>(79.59%)</td><td>26135405<br>(89.73%)</td><td>2991334<br>(10.27%)
</td></tr>
<tr><td>chronic_7</td><td>32531520</td><td>25499632<br>(78.38%)</td><td>22987035<br>(90.15%)</td><td>2512597<br>(9.85%)
</td></tr>
<tr><td>chronic_8</td><td>33730612</td><td>26356101<br>(78.14%)</td><td>23808601<br>(90.33%)</td><td>2547500<br>(9.67%)
</td></tr>
<tr><td>chronic_9</td><td>41135625</td><td>31258172<br>(75.99%)</td><td>27584413<br>(88.25%)</td><td>3673759<br>(11.75%)
</td></tr>
<tr><td>normal_1</td><td>27064427</td><td>21691950<br>(80.15%)</td><td>19818355<br>(91.36%)</td><td>1873595<br>(8.64%)
</td></tr>
<tr><td>normal_2</td><td>32961659</td><td>26471190<br>(80.31%)</td><td>24162754<br>(91.28%)</td><td>2308436<br>(8.72%)
</td></tr>
<tr><td>normal_3</td><td>33467594</td><td>27021448<br>(80.74%)</td><td>24715225<br>(91.47%)</td><td>2306223<br>(8.53%)
</td></tr>
<tr><td>normal_4</td><td>34261878</td><td>27486943<br>(80.23%)</td><td>25010950<br>(90.99%)</td><td>2475993<br>(9.01%)
</td></tr>
<tr><td>normal_5</td><td>34381801</td><td>22677156<br>(65.96%)</td><td>20427314<br>(90.08%)</td><td>2249842<br>(9.92%)</td></tr>
</tbody></table></div></div>
<p><strong>说明：</strong></p>

<p>对于转录组数据，在参考基因组选择合适并且实验不存在污染的情况下，测序序列的定位到参考基因组的</p>

<p>百分比通常会高于 70% (Total Mapped Reads)，uniquely mapped reads（单位置比对）绝大多数来自成熟的mRNA，而多位置比对（Multiple mapped）的reads以rRNA和tRNA为主，其在clean reads中所占比例通常不到10%。</p>

<p><strong>下游所有分析仅使用了uniquely mapped reads，保证分析结果的可信度。</strong></p>

       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.3 Reads在基因组不同区域的分布情况</h3>

<p><strong>方法描述：</strong></p>

<p>将Clean Reads比对到参考基因组上，统计各个区域的分布情况，统计结果如下：</p>

<p><strong>结果展示：</strong></p>

<div><p class="table_info">Table2.4.Reads distribution across reference Genomic Regions</p></div>
<div><p class="table_info">(表2.4：reads在基因组上的分布.)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample</th><th>CDS</th><th>Nc_exon</th><th>5'UTR</th><th>Intergenic</th><th>Introns</th><th>3'UTR
</th></tr>
</thead><tbody><tr><td>acute_11</td><td>9430188.35<br>(43.79%)</td><td>2225063.61<br>(10.33%)</td><td>331008.09<br>(1.54%)</td><td>989167.48<br>(4.59%)</td><td>3763800.98<br>(17.48%)</td><td>4794222.49<br>(22.26%)
</td></tr>
<tr><td>acute_12</td><td>17489970.49<br>(50.21%)</td><td>2580741.24<br>(7.41%)</td><td>707219.27<br>(2.03%)</td><td>1472708.59<br>(4.23%)</td><td>7787797.01<br>(22.36%)</td><td>4794425.40<br>(13.76%)
</td></tr>
<tr><td>acute_13</td><td>14304137.93<br>(43.67%)</td><td>2717949.66<br>(8.30%)</td><td>655499.03<br>(2.00%)</td><td>1444506.79<br>(4.41%)</td><td>8419109.41<br>(25.71%)</td><td>5210208.19<br>(15.91%)
</td></tr>
<tr><td>acute_14</td><td>12352798.61<br>(48.01%)</td><td>2017025.29<br>(7.84%)</td><td>492204.98<br>(1.91%)</td><td>1360273.24<br>(5.29%)</td><td>5185989.29<br>(20.16%)</td><td>4319552.58<br>(16.79%)
</td></tr>
<tr><td>acute_15</td><td>11076408.95<br>(44.22%)</td><td>1879236.44<br>(7.50%)</td><td>422721.41<br>(1.69%)</td><td>1141413.56<br>(4.56%)</td><td>6186297.18<br>(24.70%)</td><td>4341704.47<br>(17.33%)
</td></tr>
<tr><td>chronic_1</td><td>12178727.45<br>(53.10%)</td><td>2125332.34<br>(9.27%)</td><td>488648.43<br>(2.13%)</td><td>965743.80<br>(4.21%)</td><td>4044569.33<br>(17.64%)</td><td>3130712.65<br>(13.65%)
</td></tr>
<tr><td>chronic_6</td><td>11380778.17<br>(43.55%)</td><td>1708269.91<br>(6.54%)</td><td>523994.88<br>(2.00%)</td><td>1042956.40<br>(3.99%)</td><td>7458427.54<br>(28.54%)</td><td>4020978.09<br>(15.39%)
</td></tr>
<tr><td>chronic_7</td><td>10797147.57<br>(46.97%)</td><td>1537962.68<br>(6.69%)</td><td>474342.34<br>(2.06%)</td><td>872834.17<br>(3.80%)</td><td>5928455.19<br>(25.79%)</td><td>3376293.05<br>(14.69%)
</td></tr>
<tr><td>chronic_8</td><td>8943978.77<br>(37.57%)</td><td>1358845.74<br>(5.71%)</td><td>460834.86<br>(1.94%)</td><td>1313444.95<br>(5.52%)</td><td>8518857.74<br>(35.78%)</td><td>3212638.95<br>(13.49%)
</td></tr>
<tr><td>chronic_9</td><td>12777740.28<br>(46.32%)</td><td>2139649.00<br>(7.76%)</td><td>446667.32<br>(1.62%)</td><td>1251462.69<br>(4.54%)</td><td>6673310.68<br>(24.19%)</td><td>4295583.03<br>(15.57%)
</td></tr>
<tr><td>normal_1</td><td>8603525.31<br>(43.41%)</td><td>1493286.48<br>(7.53%)</td><td>451751.78<br>(2.28%)</td><td>995314.32<br>(5.02%)</td><td>5415782.48<br>(27.33%)</td><td>2858694.62<br>(14.42%)
</td></tr>
<tr><td>normal_2</td><td>10015240.39<br>(41.45%)</td><td>1800813.66<br>(7.45%)</td><td>530133.15<br>(2.19%)</td><td>1236262.89<br>(5.12%)</td><td>6899635.33<br>(28.55%)</td><td>3680668.58<br>(15.23%)
</td></tr>
<tr><td>normal_3</td><td>9509658.51<br>(38.48%)</td><td>1771857.27<br>(7.17%)</td><td>503819.30<br>(2.04%)</td><td>1366768.81<br>(5.53%)</td><td>8155178.17<br>(33.00%)</td><td>3407942.95<br>(13.79%)
</td></tr>
<tr><td>normal_4</td><td>10688949.03<br>(42.74%)</td><td>2115459.92<br>(8.46%)</td><td>549687.25<br>(2.20%)</td><td>1276985.19<br>(5.11%)</td><td>6482723.17<br>(25.92%)</td><td>3897145.44<br>(15.58%)
</td></tr>
<tr><td>normal_5</td><td>9086039.61<br>(44.48%)</td><td>1678689.55<br>(8.22%)</td><td>492378.29<br>(2.41%)</td><td>1009202.38<br>(4.94%)</td><td>4827173.55<br>(23.63%)</td><td>3333830.63<br>(16.32%)</td></tr>
</tbody></table></div></div>


		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g10" aria-expanded="true" aria-controls="g10">
		          region_Pie_Chart
		        </a>
		      </div>
		    </div>
		    <div id="g10" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">normal_4.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">chronic_7.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">chronic_9.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">chronic_8.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">acute_12.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">normal_3.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">chronic_6.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">normal_1.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">acute_15.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">chronic_10.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">acute_14.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">normal_5.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">acute_13.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">normal_2.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" data-lightbox="g10" data-title="Mapping_distribution.region_Pie_Chart" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/Mapping_distribution.region_Pie_Chart.png" alt="Mapping_distribution.region_Pie_Chart" /><div class="caption">acute_11.clean_fq_1230</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.4 基因组的覆盖度和特征分析</h3>

<p><strong>方法描述</strong></p>

<p>RNA-seq reads随转录单元长度的覆盖强度分析，以距5’端Tu为标准，把cDNA平均分成100份，每一份称为一个bin，求落在每个bin中的reads平均数。</p>

<p><strong>结果展示：</strong></p>

		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g11" aria-expanded="true" aria-controls="g11">
		          randCheck_mRNA
		        </a>
		      </div>
		    </div>
		    <div id="g11" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">normal_4.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">chronic_7.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">chronic_9.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">chronic_8.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">acute_12.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">normal_3.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">chronic_6.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">normal_1.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">acute_15.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">chronic_10.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">acute_14.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">normal_5.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">acute_13.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">normal_2.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" data-lightbox="g11" data-title="accepted_hits.sam.uniq.randCheck_mRNA" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/accepted_hits.sam.uniq.randCheck_mRNA.png" alt="accepted_hits.sam.uniq.randCheck_mRNA" /><div class="caption">acute_11.clean_fq_1230</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.5 reads在转录起始位点和转录终止位点附近的分布</h3>

<p><strong>方法描述</strong></p>

<p>分别以转录起始位点/转录终止位点为原点，统计其上下游1kb范围内reads的分布情况。</p>

<p><strong>结果展示：</strong></p>

		<div class="row">

	  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
    <div class="panel-heading" role="tab" id="headingOne">
      <div class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion" href="#g12" aria-expanded="true" aria-controls="g12">
          Collapsible Group Item #1
        </a>
      </div>
    </div>
    <div id="g12" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
      <div class="panel-body">
		<div class="row">
	</div>
		</div>
      </div>
    </div>
    
  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.6 reads在起始密码子和终止密码子附近的分布</h3>

<p><strong>方法描述</strong></p>

<p>分别以起始密码子/终止密码子为原点，统计其上下游1kb范围内reads的分布情况。</p>

<p><strong>结果展示：</strong></p>

		<div class="row">

	  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
    <div class="panel-heading" role="tab" id="headingOne">
      <div class="panel-title">
        <a data-toggle="collapse" data-parent="#accordion" href="#g13" aria-expanded="true" aria-controls="g13">
          Collapsible Group Item #1
        </a>
      </div>
    </div>
    <div id="g13" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
      <div class="panel-body">
		<div class="row">
	</div>
		</div>
      </div>
    </div>
    
  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h3>3.2.7 基因表达情况</h3>

<p><strong>方法描述：</strong></p>

<p>RPKM (Reads per kilo base of a gene per million reads)，表示每百万 reads 中来自于某基因每千碱基长度的 reads 数，表征基因的表达丰度。通过对mRNA长度和测序深度进行均一化(RPKM)，使不同测序样本之间的表达丰度具有可比性，消除了因mRNA长度和不同样本之间测序丰度差异带来的偏差。</p>

<p>在这部分分析中，我们对本项目检出的基因中RPKM≥0和RPKM≥10的基因数进行统计，统计结果如下：</p>

<p><strong>结果展示：</strong></p>

<div><p class="table_info">Table2.5.Mapped normalized reads on expressed genes in reference genome</p></div>
<div><p class="table_info">(表2.5：基因表达水平统计.)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>Sample</th><th>Total genes in genome</th><th>Expressed genes(RPKM>0)(%) a</th><th>Expressed genes(RPKM>=10)(%) b</th></tr>
</thead><tbody><tr><td>acute11</td><td>28518</td><td>19060<br>(66.83%)</td><td>4450<br>(23.35%)</td></tr>
<tr><td>acute12</td><td>28518</td><td>18611<br>(65.26%)</td><td>5124<br>(27.53%)</td></tr>
<tr><td>acute13</td><td>28518</td><td>19758<br>(69.28%)</td><td>5504<br>(27.86%)</td></tr>
<tr><td>acute14</td><td>28518</td><td>18550<br>(65.05%)</td><td>4530<br>(24.42%)</td></tr>
<tr><td>acute15</td><td>28518</td><td>19401<br>(68.03%)</td><td>5045<br>(26.00%)</td></tr>
<tr><td>chronic10</td><td>28518</td><td>17028<br>(59.71%)</td><td>3467<br>(20.36%)</td></tr>
<tr><td>chronic6</td><td>28518</td><td>18620<br>(65.29%)</td><td>5266<br>(28.28%)</td></tr>
<tr><td>chronic7</td><td>28518</td><td>18682<br>(65.51%)</td><td>5177<br>(27.71%)</td></tr>
<tr><td>chronic8</td><td>28518</td><td>19050<br>(66.80%)</td><td>4821<br>(25.31%)</td></tr>
<tr><td>chronic9</td><td>28518</td><td>18442<br>(64.67%)</td><td>4023<br>(21.81%)</td></tr>
<tr><td>normal1</td><td>28518</td><td>18436<br>(64.65%)</td><td>5415<br>(29.37%)</td></tr>
<tr><td>normal2</td><td>28518</td><td>18908<br>(66.30%)</td><td>5570<br>(29.46%)</td></tr>
<tr><td>normal3</td><td>28518</td><td>18857<br>(66.12%)</td><td>5486<br>(29.09%)</td></tr>
<tr><td>normal4</td><td>28518</td><td>18732<br>(65.68%)</td><td>5382<br>(28.73%)</td></tr>
<tr><td>normal5</td><td>28518</td><td>19133<br>(67.09%)</td><td>5493<br>(28.71%)</td></tr>
<tr><td></td><td></td><td>a 表达的gene数占总参考<br>基因组中gene总数的比例</td><td>b RPKM值大于等于10的gene数<br>占RPKM值大于0的gene的比例</td></tr>
</tbody></table></div></div>


		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g14" aria-expanded="true" aria-controls="g14">
		          accepted_hits_RPKM
		        </a>
		      </div>
		    </div>
		    <div id="g14" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_4.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">normal_4.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_7.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">chronic_7.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_9.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">chronic_9.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_8.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">chronic_8.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_12.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">acute_12.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_3.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">normal_3.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_6.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">chronic_6.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_1.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">normal_1.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_15.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">acute_15.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/chronic_10.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">chronic_10.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_14.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">acute_14.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_5.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">normal_5.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_13.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">acute_13.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/normal_2.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">normal_2.clean_fq_1230</div></a></div>
				<div class="col-xs-3 col-md-3"><a href="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/accepted_hits_RPKM.png" data-lightbox="g14" data-title="accepted_hits_RPKM" class="thumbnail"><img src="RNA-seq/Sup2_Mapping/acute_11.clean_fq_1230/accepted_hits_RPKM.png" alt="accepted_hits_RPKM" /><div class="caption">acute_11.clean_fq_1230</div></a></div>
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h1>Part Ⅳ转录组差异表达分析方案与结果</h1>

<h2>4.1 标准DEG分析流程</h2>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="assets/public/DEG_workflow.jpg" data-lightbox="g15" data-title="Figure 4.1．DEG workflow" class="thumbnail"><img src="assets/public/DEG_workflow.jpg" alt="Figure 4.1．DEG workflow" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h2>4.2 样本相关性分析</h2>

<p>首先我们进行两样本之间的相关性分析，即检查不同样品之间同一基因表达的相关性。如果大部分基因的表达量相似，相关系数高，说明数据的均一化程度高。如果相关系数低，则暗示基因表达模式存在显著差异，或者数据质量有问题。</p>

<p><strong>结果展示：</strong></p>

		<div class="row">
			
			  <div class="panel panel-default col-xs-10 col-md-10 center-block" style="float:none;padding:0px;">
		    <div class="panel-heading" role="tab" id="headingOne">
		      <div class="panel-title">
		        <a data-toggle="collapse" data-parent="#accordion" href="#g16" aria-expanded="true" aria-controls="g16">
		          lm
		        </a>
		      </div>
		    </div>
		    <div id="g16" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="headingOne">
		  <div class="panel-body">
		<div class="row">
			</div>
				</div>
		      </div>
		    </div>
		    
		  </div>
<h2>4.3 样本聚类分析</h2>

<p><strong>方法描述</strong></p>

<p>在这部分分析中，我们根据各样本基因表达情况进行样本相关性分析，根据样本相关性系数进行样本间聚类分析。</p>

<p><strong>结果展示：</strong></p>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="RNA-seq/Sup3_Correlation_analysis/Sample_correlation.jpg" data-lightbox="g17" data-title="Figure 4.2．Sample_correlation" class="thumbnail"><img src="RNA-seq/Sup3_Correlation_analysis/Sample_correlation.jpg" alt="Figure 4.2．Sample_correlation" /></a></div></div>


       <div class="tb-block noprint">
            <div class="blblock"></div>
            <div class="brblock"></div>
        </div>
    </div>


    <!-- new page -->
    <div class="page noprintborder">
        <div class="tb-block noprint">
            <div class="tlblock"></div>
            <div class="trblock"></div>
        </div>
        <div class="pageheader">
            <img src="assets/img/logo.png">
            <span class="headerword">ABLIFE TRANSLATES SEQUENCE TO SCIENCE AND INDUSTRY</span>
            <hr class="hr0"/>
            <hr class="hr1"/>
        </div>

<h2>4.4 差异表达分析</h2>

<p><strong>方法描述：</strong></p>

<p>对样本间进行比较，获得差异表达基因（differentially expressed gene, DEG）。</p>

<h3>4.4.1 标准DEG</h3>

<p><strong>方法描述</strong></p>

<p>在做基因的差异表达分析时，我们选用的软件是 edgeR，它是一款专门针对基因表达量做差异表达分析的软件，用到的数学模型是负二项分布模型；edgeR 可以分析两个或者多个样品间的差异表达，分析结果以fold change 和 P-value 来判断一个基因是否差异表达。</p>

<p><strong>两个重要参数：</strong></p>

<p>1) FC：fold change：表达变化的绝对比值。</p>

<p>2) P-value：表达变化的统计学可信度。</p>

<p><strong>DEG standard:</strong></p>

<p>fold change≥2 or ≤0.5, p-value≤0.01;</p>

<p>CPM 代表 count per million，指的是每百万reads 中该基因的reads 数，先分别计算同一基因在相比较的两个样本中的 CPM 值，再取平均值，然后再以2为底取log，就得到了logCPM。 </p>

<p><strong>结果展示（fold change≥2或≤0.5，P-value≤0.01) ：</strong></p>

0

<div><p class="table_info">Table4.1. Total number of differentially expressed genes</p></div>
<div><p class="table_info">(表4.1：差异表达基因统计)</p></div>
	<div class="row"><div class="table-responsive col-xs-8 col-md-8 center-block" style="float:none;"><table class="table table-hover table-striped"><thead>
<tr><th>DEG sample</th><th>Total genes</th><th>Up regulated gene NO.</th><th>Down regulated gene NO.</th><th>Total
</th></tr>
</thead><tbody><tr><td>Acute_vs_Chronic</td><td>28518</td><td>1934</td><td>1067</td><td>3001
</td></tr>
<tr><td>Acute_vs_Normal</td><td>28518</td><td>3551</td><td>1706</td><td>5257
</td></tr>
<tr><td>Chronic_sv_Normal</td><td>28518</td><td>3943</td><td>3266</td><td>7209</td></tr>
</tbody></table></div></div>
<h2>4.4.2 差异基因维恩图</h2>

<p>维恩图主要展示每组比较中差异基因的个数，以及多组比较中共同的差异基因个数。</p>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="RNA-seq/Sup4_DEG_results/DEG Venn.jpg" data-lightbox="g18" data-title="Figure 4.3．DEG Venn" class="thumbnail"><img src="RNA-seq/Sup4_DEG_results/DEG Venn.jpg" alt="Figure 4.3．DEG Venn" /></a></div></div>
<h2>4.4.3 差异基因聚类图</h2>

<p><strong>方法描述</strong></p>

<p>差异基因聚类分析用于揭示不同实验条件下基因表达模式的差异。</p>

	<div class="row"><div class="col-xs-7 col-md-7 center-block" style="float:none;"><a href="RNA-seq/Sup4_DEG_results/DEG heatmap.jpg" data-lightbox="g19" data-title="Figure 4.4．DEG heatmap" class="thumbnail"><img src="RNA-seq/Sup4_DEG_results/DEG heatmap.jpg" alt="Figure 4.4．DEG heatmap" /></a></div></div>
<h2>4.5 差异基因Gene Ontology 富集分析</h2>

<p><strong>方法描述</strong></p>

<ol>
<li><p>利用blast将参考基因组的基因序列比对到Gene Ontology数据库，进行GO注释；</p></li>
<li><p>提取比对结果，作为背景即background；</p></li>
<li><p>根据DEG分析结果，统计每个DEG所在的GO Term，根据每个Term的基因数目，以及背景中此Term的基因数目，用Fisher Exact Test分析每个Term的显著性。</p></li>
</ol>

<p><strong>结果展示：</strong></p>

<h2>4.6 差异基因KEGG Pathway富集分析</h2>

<p><strong>方法描述</strong></p>

<ol>
<li><p>利用blast将参考基因组的基因序列比对到KEGG数据库，进行KEGG注释；</p></li>
<li><p>提取比对结果，作为背景即background；</p></li>
<li><p>根据DEG分析结果，统计每个DEG所在的KEGG pathway，根据每个pathway的基因数目，以及背景中此pathway的基因数目，用Fisher Exact Test分析每个pathway的显著性。</p></li>
</ol>

<p><strong>结果展示：</strong></p>




    </div>
    <!-- /pageend -->


</div>
<!-- /container -->


<div id="modal noprint"></div>


<div class="menu noprint">
    <div id="gotop" class="item">
        <a class="link icon_top"></a>
        <a class="item_content">Top</a>
    </div>
    <div id="print" class="item">
        <a class="link icon_print"></a>
        <a class="item_content">打印</a>
    </div>
    <div id="expand" class="item">
        <a class="link icon_expand"></a>
        <a class="item_content">版式</a>
    </div>
    <div class="item">
        <a href="#modal" class="link icon_menu second"></a>
        <a href="
    </div>
</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="assets/js/jquery.min.js"></script>
<script src="assets/js/jquery-ui.js"></script>
<script src="assets/js/bootstrap.js"></script>

<!-- IE10 viewport hack for Surface/desktop Windows 8 bug -->
<script src="assets/js/ie10-viewport-bug-workaround.js"></script>

<!-- other JavaScript
================================================== -->
<!-- lightbox js -->
<script src="assets/js/lightbox.js"></script>

<!-- ztree js -->
<script type="text/javascript" src="assets/js/jquery.ztree.all-3.5.js"></script>
<script type="text/javascript" src="assets/js/ztree_toc.js"></script>

<!-- pageslide js -->
<script src="assets/js/jquery.pageslide.js"></script>
<script>
    /* Slide to the left, and make it model (you'll have to call $.pageslide.close() to close) */
    $(".second").pageslide({direction: "right", modal: true});
</script>

<!-- tools js -->
<script type="text/javascript">

    $(function () {



        //点击回到顶部的元素
        $("#gotop").click(function (e) {
            //以1秒的间隔返回顶部
            $('body,html').animate({scrollTop: 0}, 300);
        });

        $("#print").click(function (e) {
            //以1秒的间隔返回顶部
            $.pageslide.close();
            window.print();
        });

        $("#expand").click(function (e) {
            if ($('.tb-block').is(':visible')) {
                $('.tb-block').css("display", "none");
                $('.pageheader').css("display", "block");
                $('.pagefooter').css("display", "block");
                $('.page').css("margin", "30px 0");
                $('#container').css("padding", "30px 0");
                $('.icon_expand').css("background", "transparent url(assets/img/collapse.png) no-repeat top left");
            } else {
                $('.tb-block').css("display", "block");
                $('.page').css("margin", "0");
                $('.pageheader').css("display", "none");
                $('.pagefooter').css("display", "none");
                $('#container').css("padding", "30px 0");
                $('.icon_expand').css("background", "transparent url(assets/img/expand.png) no-repeat top left");
            }

        });
    });

    $(".item").mouseover(function (e) {
        $(this).css("opacity", "1");
    });
    $(".item").mouseout(function (e) {
        $(this).css("opacity", "0.7");
    });
    $('.item').hover(
            function () {
                var $this = $(this);
                expand($this);
            },
            function () {
                var $this = $(this);
                collapse($this);
            }
    );
    function expand($elem) {
        $elem.stop().animate({width: '120px'}, 500)
                .find('.item_content').fadeIn(200, function () {
                    $(this).find('p').stop(true, true).fadeIn(300);
                });
    }
    function collapse($elem) {
        $elem.stop().animate({width: '52px'}, 500)
                .find('.item_content').stop(true, true).fadeOut().find('p').stop(true, true).fadeOut();
    }
    $(function () {
      $('[data-toggle="tooltip"]').tooltip();
    });
</script>
</body>
</html>
